pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,452 @@
+		return err
+	}
+
+	err = t.sendPayload(dest, out)
+	if err != nil {
+		t.reconnect(ctx, dest)
+		return errors.Wrapf(err, "%s failed to send data to %s", t.ownID, dest)
+	}
+
+	return nil
+}
+
+func (t *Transport) ImplementsModule() {}
+
+func (t *Transport) EventsOut() <-chan *events.EventList {
+	return t.incomingMessages
+}
+
+func (t *Transport) ApplyEvents(ctx context.Context, eventList *events.EventList) error {
+	iter := eventList.Iterator()
+	for event := iter.Next(); event != nil; event = iter.Next() {
+
+		switch e := event.Type.(type) {
+		case *eventpb.Event_Init:
+			// no actions on init
+		case *eventpb.Event_SendMessage:
+			for _, destID := range e.SendMessage.Destinations {
+				if types.NodeID(destID) == t.ownID {
+					// Send message to myself bypassing the network.
+					// The sending must be done in its own goroutine in case writing to gt.incomingMessages blocks.
+					// (Processing of input events must be non-blocking.)
+					receivedEvent := events.MessageReceived(
+						types.ModuleID(e.SendMessage.Msg.DestModule),
+						t.ownID,
+						e.SendMessage.Msg,
+					)
+					go func() {
+						select {
+						case t.incomingMessages <- events.ListOf(receivedEvent):
+						case <-ctx.Done():
+						}
+					}()
+				} else {
+					// Send message to another node.
+					if err := t.Send(ctx, types.NodeID(destID), e.SendMessage.Msg); err != nil {
+						t.logger.Log(logging.LevelError, "failed to send a message", "err", err)
+					}
+				}
+			}
+		default:
+			return fmt.Errorf("unexpected event: %T", event.Type)
+		}
+	}
+
+	return nil
+}
+
+func (t *Transport) connect(ctx context.Context, nodes map[types.NodeID]types.NodeAddress) {
+	t.connWg.Add(1)
+
+	go func() {
+		defer t.connWg.Done()
+
+		t.connsLock.Lock()
+		defer t.connsLock.Unlock()
+
+		t.logger.Log(logging.LevelDebug, "started connecting nodes", "src", t.ownID)
+		defer t.logger.Log(logging.LevelDebug, "finished connecting nodes", "src", t.ownID)
+
+		wg := &sync.WaitGroup{}
+		wg.Add(len(nodes))
+
+		for nodeID := range nodes {
+			// Do not establish a real connection with own node.
+			if nodeID == t.ownID {
+				wg.Done()
+				continue
+			}
+
+			go t.connectToNodeWithoutLock(ctx, nodeID, wg)
+		}
+		wg.Wait()
+	}()
+}
+
+func (t *Transport) reconnect(ctx context.Context, node types.NodeID) {
+	info, found := t.getNodeAddrInfo(node)
+	if !found {
+		t.logger.Log(logging.LevelError, "failed to get node address", "src", t.ownID, "dst", node)
+		return
+	}
+	if err := t.host.Network().ClosePeer(info.ID); err != nil {
+		t.logger.Log(logging.LevelError, "could not close the connection to node", "src", t.ownID, "dst", node, "err", err)
+	}
+
+	t.connWg.Add(1)
+	go func() {
+		defer t.connWg.Done()
+
+		t.connsLock.Lock()
+		defer t.connsLock.Unlock()
+
+		t.connWg.Add(1)
+		t.connectToNodeWithoutLock(ctx, node, t.connWg)
+	}()
+}
+
+func (t *Transport) connectToNodeWithoutLock(ctx context.Context, nodeID types.NodeID, wg *sync.WaitGroup) {
+	defer wg.Done()
```

Comment:
```suggestion
	defer wg.Done()
```

This is not needed anymore right? You handle the waiting groups in the functions where they are called, mainly `reconnect` and `connect`